jujugui/static/gui/src/test/test_app.js

```diff
@@ -1954,7 +1954,396 @@ describe('App', function() {
+        controllerSocketTemplate: '/api',
+        viewContainer: container,
+        jujuCoreVersion: '2.0.0'
+      });
+    });
+
+    afterEach(function() {
+      app.destroy();
+    });
+
+    it('properly handles a resolved entity promise', done => {
+      const userState = {user: 'hatch'};
+      const entityPromise = Promise.reject(userState);
+      app.controllerAPI.userIsAuthenticated = true;
+      app._listAndSwitchModel = args => {
+        assert.deepEqual(args, userState);
+        done();
+      };
+      app._disambiguateUserState(entityPromise);
+    });
+
+    it('properly handles a rejected entity promise', done => {
+      const userState = 'hatch';
+      const entityPromise = Promise.resolve(userState);
+      app.maskVisibility = sinon.stub();
+      app.state.changeState = state => {
+        assert.deepEqual(app.maskVisibility.args[0], [false]);
+        assert.deepEqual({
+          store: 'u/hatch',
+          user: null
+        }, state);
+        done();
+      };
+      app._disambiguateUserState(entityPromise);
+    });
+  });
+
+  describe('_listAndSwitchModel', function() {
+    let app, modelList, juju, Y;
+
+    before(function(done) {
+      Y = YUI(GlobalConfig).use(['juju-gui', 'juju-tests-utils'], function(Y) {
+        juju = juju = Y.namespace('juju');
+        done();
+      });
+    });
+
+    beforeEach(function() {
+      app = new Y.juju.App({
+        baseUrl: 'http://example.com/',
+        controllerAPI: new juju.ControllerAPI({
+          conn: new testUtils.SocketStub()
+        }),
+        env: new juju.environments.GoEnvironment({
+          conn: new testUtils.SocketStub(),
+          ecs: new juju.EnvironmentChangeSet(),
+          user: 'user',
+          password: 'password'
+        }),
+        socketTemplate: '/model/$uuid/api',
+        controllerSocketTemplate: '/api',
+        viewContainer: container,
+        jujuCoreVersion: '2.0.0'
+      });
+      modelList = [{
+        'id':'fe9a2845-4829-4d61-8653-248b7052204e',
+        'name':'latta',
+        'series':'xenial',
+        'provider':'gce',
+        'uuid':'fe9a2845-4829-4d61-8653-248b7052204e',
+        'controllerUUID':'a030379a-940f-4760-8fcf-3062b41a04e7',
+        'owner':'frankban@external',
+        'life':'alive',
+        'isAlive':true,
+        'isController':false,
+        'lastConnection':null
+      },{
+        'id':'509f6e4c-4da4-49c8-8f18-537c33b4d3a0',
+        'name':'jujugui-org',
+        'series':'xenial',
+        'provider':'gce',
+        'uuid':'509f6e4c-4da4-49c8-8f18-537c33b4d3a0',
+        'controllerUUID':'a030379a-940f-4760-8fcf-3062b41a04e7',
+        'owner':'uros-jovanovic@external',
+        'life':'alive',
+        'isAlive':true,
+        'isController':false,
+        'lastConnection':null
+      }];
+    });
+
+    afterEach(function() {
+      app.destroy();
+    });
+
+    it('switches to a supplied model path', () => {
+      app.maskVisibility = sinon.stub();
```

Comment:
ISTM that these initial lines are the same for each tests, so I guess they can be moved to beforeEach:
```
app.maskVisibility = sinon.stub();
app.state.changeState = sinon.stub();
app._getAuth = sinon.stub().returns({rootUserName: 'pug'});
app.controllerAPI.listModelsWithInfo = function(callback) {
  callback.call(app, null, modelList);
};
```